<div style="border-radius: 20px; border: 2px solid #64B5F6; padding: 15px; background-color: #000000; text-align: center; box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.4), 0px 6px 20px rgba(0, 0, 0, 0.19); transform: perspective(1000px) rotateX(5deg) rotateY(-5deg); transition: transform 0.5s ease-in-out;">
    <h1 style="color: #1976D2; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.4); font-weight: bold; margin-bottom: 10px; font-size: 32px;">
        Asalam alaikum warahmatullah wabarakatu!
    </h1>
    <p style="color: #1976D2; font-size: 18px; margin: 10px 0;">
        I am Mustafa Shoukat, deeply engaged in the field of AI. Join me as we explore cutting-edge concepts and techniques, advancing our skills together. In this notebook my aims to examine various Hugging Face models, with the objective of increasing my confidence in applying and customizing pretrained models for different real world tasks.
    </p>
    <p style="color: #43A047; font-size: 16px; font-style: italic; margin: 10px 0;">
        "The power of community lies in our combined efforts. United, we can reach heights we never imagined."
    </p>
    <h2 style="color: #1976D2; margin-top: 15px; font-size: 28px;">Contact Information</h2>
    <table style="width: 100%; margin-top: 15px; border-collapse: collapse;">
        <tr>
            <th style="color: #1976D2; font-size: 18px; padding: 8px; border-bottom: 2px solid #64B5F6;">Name</th>
            <th style="color: #1976D2; font-size: 18px; padding: 8px; border-bottom: 2px solid #64B5F6;">Email</th>
            <th style="color: #1976D2; font-size: 18px; padding: 8px; border-bottom: 2px solid #64B5F6;">LinkedIn</th>
            <th style="color: #1976D2; font-size: 18px; padding: 8px; border-bottom: 2px solid #64B5F6;">GitHub</th>
            <th style="color: #1976D2; font-size: 18px; padding: 8px; border-bottom: 2px solid #64B5F6;">Kaggle</th>
        </tr>
        <tr>
            <td style="font-size: 16px; padding: 8px;">Mustafa Shoukat</td>
            <td style="font-size: 16px; padding: 8px;">mustafashoukat.ai@gmail.com</td>
            <td style="font-size: 16px; padding: 8px;">
                <a href="https://www.linkedin.com/in/mustafashoukat/" target="_blank">
                    <img src="https://img.shields.io/badge/LinkedIn-0e76a8.svg?style=for-the-badge&logo=LinkedIn&logoColor=white" alt="LinkedIn Badge" style="border-radius: 5px;">
                </a>
            </td>
            <td style="font-size: 16px; padding: 8px;">
                <a href="https://github.com/Mustafa-Shoukat1" target="_blank">
                    <img src="https://img.shields.io/badge/GitHub-171515.svg?style=for-the-badge&logo=GitHub&logoColor=white" alt="GitHub Badge" style="border-radius: 5px;">
                </a>
            </td>
            <td style="font-size: 16px; padding: 8px;">
                <a href="https://www.kaggle.com/mustafashoukat" target="_blank">
                    <img src="https://img.shields.io/badge/Kaggle-20beff.svg?style=for-the-badge&logo=Kaggle&logoColor=white" alt="Kaggle Badge" style="border-radius: 5px; margin: 0 5px;">
                </a>
            </td>
        </tr>
    </table>
</div>




<div style="border-radius: 20px; border: 2px solid #64B5F6; padding: 15px; background-color: #000000; color: #FFFFFF; text-align: center; box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.4), 0px 6px 20px rgba(0, 0, 0, 0.19); transform: perspective(1000px) rotateX(5deg) rotateY(-5deg); transition: transform 0.5s ease-in-out;">
    <h1 style="color: #1976D2; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.4); font-weight: bold; margin-bottom: 10px; font-size: 32px;">
        Purpose:
    </h1>
    <p style="font-size: 18px; margin: 10px 0;">
        Introduction to Hugging Face Transformers, Datasets and Models
    </p>
    <p style="font-size: 18px; margin: 10px 0;">
        This notebook provides a practical exploration of Hugging Face's ecosystem, focusing on datasets, models, and transformers. Hugging Face has become a pivotal platform for Advance Data Science, offering powerful tools and pre-trained models that streamline the development of various applications.
    </p>
    <h2 style="font-weight: bold; margin-bottom: 10px; font-size: 24px;">
        Transformers and Pipelines
    </h2>
    <p style="font-size: 18px; margin: 10px 0;">
        Hugging Face Transformers enable a wide range of NLP tasks through easy-to-use pipelines. Key functionalities include:
    </p>
    <ul style="font-size: 18px; margin: 10px 0; text-align: left;">
        <li><strong>Named Entity Recognition (NER):</strong> Identifying entities like names, dates, and locations in text.</li>
        <li><strong>Question Answering:</strong> Providing answers to questions posed in natural language.</li>
        <li><strong>Summarization:</strong> Generating concise summaries of input text.</li>
        <li><strong>Translation:</strong> Translating text between different languages.</li>
        <li><strong>Text Classification:</strong> Categorizing text into predefined classes.</li>
    </ul>
    <h2 style="font-weight: bold; margin-bottom: 10px; font-size: 24px;">
        Datasets and Tokenization
    </h2>
    <p style="font-size: 18px; margin: 10px 0;">
        The platform includes access to various datasets and efficient tokenization mechanisms, essential for preparing data for model training and evaluation.
    </p>
    <h2 style="font-weight: bold; margin-bottom: 10px; font-size: 24px;">
        Hugging Face Models
    </h2>
    <p style="font-size: 18px; margin: 10px 0;">
        Hugging Face hosts a repository of state-of-the-art models, including:
    </p>
    <ul style="font-size: 18px; margin: 10px 0; text-align: left;">
        <li><strong>Mixtral-8x22B-v0.1:</strong> A Mixture of Experts model known for its high performance.</li>
    </ul>
    <h2 style="font-weight: bold; margin-bottom: 10px; font-size: 24px;">
        Mixture of Experts Model
    </h2>
    <p style="font-size: 18px; margin: 10px 0;">
        Exploring advanced models like the Mixtral-8x22B-v0.1, demonstrating the capabilities of complex architectures in NLP tasks.
    </p>
    <h2 style="font-weight: bold; margin-bottom: 10px; font-size: 24px;">
        Evaluation Metrics
    </h2>
    <p style="font-size: 18px; margin: 10px 0;">
        Evaluation of models includes metrics such as:
    </p>
    <ul style="font-size: 18px; margin: 10px 0; text-align: left;">
        <li><strong>Bert Score:</strong> Evaluating the quality of text generation.</li>
        <li><strong>BLEU (Bilingual Evaluation Understudy):</strong> Assessing machine-translated text.</li>
        <li><strong>GLUE (General Language Understanding Evaluation):</strong> A benchmark for evaluating language understanding models.</li>
        <li><strong>Perplexity:</strong> Measuring how well a probability model predicts a sample.</li>
    </ul>
    <p style="font-size: 18px; margin: 10px 0;">
        This notebook aims to provide hands-on experience with these tools and concepts, empowering developers to leverage Hugging Face for advanced NLP applications.
    </p>
</div>


<div style="display: flex; justify-content: space-between;">
    <img src="https://hdqwalls.com/wallpapers/transformers-age-of-extinction-optimus-prime.jpg" alt="Optimus Prime" width="400" height="400">
    <img src="https://th.bing.com/th/id/OIP.jeglzQmJMe6Z8fp27OvOpwHaD3?rs=1&pid=ImgDetMain" alt="Hugging Face" width="400" height="400">
</div>

In [1]:
!pip install transformers
!pip install datasets
!pip install sentencepiece

In [2]:
text = """ code geass is one of those series that everybody recomends and once you watch it you know why.

Mechas, war, romance, comedy, rivality, unexpected twsits, explotions, mindgames and more mechas! what else do you want!!! xD it totally deserver its place as one of the best animes ever made.

the only aspect i see as a flaw is the art with a 9/10 the people seem a little bit deformed or too skinny but once you get use to it is great!.The user wants freedom.

and it has by far the best story i have ever seen you just get hooked to it from the second episode and cant stop until finishing it.

almost perfect serie, recomended to anyone who is a human beeing (nevermind, my dog liked it) """

# Natural Language Processing with Transformers, Datasets, and SentencePiece

## Overview

This notebook demonstrates the use of state-of-the-art Natural Language Processing (NLP) tools and libraries to build and fine-tune models for various NLP tasks. We will leverage Hugging Face's `Transformers` and `Datasets` libraries alongside Google's `SentencePiece` for tokenization and text normalization.

## Libraries and Tools

### Transformers

Developed by [Hugging Face](https://huggingface.co/), the `Transformers` library provides a variety of pre-trained models that have revolutionized NLP. These models, such as BERT, GPT, and RoBERTa, are designed for tasks like:

- Text classification
- Question answering
- Text generation
- Language translation

Key features:
- Easy loading and fine-tuning of pre-trained models.
- Interfaces for both TensorFlow and PyTorch.

### Datasets

Also developed by [Hugging Face](https://huggingface.co/), the `Datasets` library offers high-quality datasets for training and evaluating NLP models. It supports tasks including:

- Text classification
- Named entity recognition (NER)
- Machine translation

Key features:
- Unified interface for accessing and manipulating datasets.
- Functions for downloading, splitting, preprocessing, and loading datasets.

### SentencePiece

Developed by Google, `SentencePiece` is a tokenizer and text normalizer that segments text into subword units. This approach is particularly useful for handling out-of-vocabulary words and reducing vocabulary size.

Key features:
- Uses a unigram language model to learn tokenization from text.
- Breaks new text into subwords based on learned patterns.
- Compatible with NLP frameworks like TensorFlow and PyTorch.

## Workflow

1. **Loading Pre-trained Models and Datasets:**
   - Utilize the `Transformers` library to load state-of-the-art pre-trained models.
   - Access and manipulate datasets using the `Datasets` library.

2. **Tokenization and Text Normalization:**
   - Apply `SentencePiece` for effective tokenization, which helps in managing out-of-vocabulary words and reduces the vocabulary size.

3. **Fine-tuning and Evaluation:**
   - Fine-tune the pre-trained models on specific NLP tasks.
   - Evaluate model performance using the provided datasets.

## Conclusion

By integrating `Transformers`, `Datasets`, and `SentencePiece`, we can efficiently build and fine-tune robust NLP models capable of handling a variety of tasks. This notebook serves as a comprehensive guide to harnessing the power of these advanced tools in your NLP projects.


In [3]:
import datasets
import huggingface_hub
import matplotlib.pyplot as plt
import transformers

- huggingface_hub democratizes access to cutting-edge NLP models and datasets by providing a centralized repository where users can easily discover and download resources for their projects.



- By providing a centralized hub for hosting models and datasets, huggingface_hub streamlines the process of sharing and accessing resources, thereby promoting collaboration, reproducibility, and innovation in the field of NLP.

In [4]:
# Text Classification
from transformers import pipeline
classifier = pipeline('text-classification')

2024-06-26 15:23:33.529465: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-06-26 15:23:33.529587: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-06-26 15:23:33.642007: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/opt/conda/lib/python3.10/site-packages/huggingface_hub/file_downloa

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

# **Transformers Pipeline**

The core architecture behind the text classification pipeline in the Hugging Face Transformers library typically involves a pre-trained transformer model fine-tuned on a large corpus of text data for the specific task of text classification. Here's how it works, illustrated with an example:

## **Loading the Model:**

When you create a text classification pipeline using `pipeline('text-classification')`, the Hugging Face library automatically downloads the pre-trained model associated with text classification tasks. This model is typically a transformer-based architecture like BERT, RoBERTa, or DistilBERT that has been fine-tuned on a dataset for text classification.

## **Tokenization:**

The input text provided to the pipeline is tokenized using the same tokenizer that was used during the pre-training of the model. This tokenizer converts the input text into a sequence of tokens that can be processed by the transformer model.

## **Encoding:**

The tokenized input text is then encoded into numerical representations suitable for input to the transformer model. This encoding typically involves mapping each token to its corresponding index in the model's vocabulary and converting the input sequence into numerical vectors.

## **Model Inference:**

The encoded input sequence is fed into the pre-trained transformer model. The model processes the input sequence through multiple layers of self-attention mechanisms and feed-forward neural networks to extract contextual representations of the text.

## **Classification Head:**

After processing the input text through the transformer layers, a classification head is attached to the model. This classification head consists of one or more dense layers that take the final contextual representation of the input text as input and produce predictions for the classification task.

- For example, in binary classification tasks, the classification head may consist of a single neuron with a sigmoid activation function that outputs a probability score indicating the likelihood of the input text belonging to a particular class.
- In multi-class classification tasks, the classification head may consist of multiple neurons corresponding to each class, with a softmax activation function applied to produce a probability distribution over the classes.

## **Prediction:**

Finally, the model outputs the predicted class label or probability distribution over the classes for the input text. This prediction is returned as the output of the text classification pipeline.


**config.json:**

- This file contains the configuration parameters of the model, including details such as the model architecture, hyperparameters, tokenizer settings, and any other model-specific configurations. It provides essential information needed to instantiate and use the model in your code.

        {
          "attention_probs_dropout_prob": 0.1,    ** > This parameter determines the dropout probability for attention weights. Dropout is a regularization technique used during training to prevent overfitting by randomly setting some activations to zero.**
          "hidden_dropout_prob": 0.1,   Similar to attention_probs_dropout_prob, this parameter specifies the dropout probability for the hidden layers. It controls the dropout applied to the outputs of each layer in the model.
          "hidden_size": 768,   This parameter defines the dimensionality of the hidden layers in the transformer model. It represents the number of neurons in each layer's feedforward network.
          "num_attention_heads": 12,
          "num_hidden_layers": 12,   The total number of layers in the transformer model, including both encoder and decoder layers. Each layer consists of sublayers like self-attention, feedforward, and layer normalization.
          "type_vocab_size": 2,
          "vocab_size": 30522,
          ...
        }


**model.safetensors**


- safetensors is a safe and fast file format for storing and loading tensors. Typically, PyTorch model weights are saved or pickled into a .bin file with Python’s pickle utility. However, pickle is not secure and pickled files may contain malicious code that can be executed. safetensors is a secure alternative to pickle, making it ideal for sharing model weights.<br><br>


          # Shopping cart data
          shopping_cart = {
              "items": [
                  {"name": "Laptop", "quantity": 1, "price": 999.99},
                  {"name": "Headphones", "quantity": 2, "price": 49.99},
                  {"name": "Mouse", "quantity": 1, "price": 19.99}
              ],
              "total": 1119.96
          }


          import json

          # Serialize the shopping cart to JSON
          serialized_cart = json.dumps(shopping_cart)

          # Save the serialized data to a file
          with open("shopping_cart.json", "w") as file:
              file.write(serialized_cart)



                        # Read the serialized data from the file
          with open("shopping_cart.json", "r") as file:
              serialized_cart = file.read()

          # Deserialize the JSON data into a Python dictionary
          restored_cart = json.loads(serialized_cart)

          print(restored_cart)



**pytorch_model.bin:**

- This file contains the actual weights and parameters of the BERT model, learned during pre-training or fine-tuning.
Example content: Serialized tensor data representing the weights of each layer, attention matrices, and other parameters of the model.

- Serialization: Serialization is the process of converting complex data structures, such as tensors, into a format that can be easily stored, transmitted, or reconstructed later. Serialization typically involves converting the data into a byte stream or a string format that preserves the structure and content of the original data.<br><br>

        import pickle

        # Example data to serialize
        data = {'name': 'John', 'age': 30, 'city': 'New York'}

        # Serialize data using pickle and save to a file
        with open('data.pkl', 'wb') as file:
            pickle.dump(data, file)

        # Deserialize data from the file
        with open('data.pkl', 'rb') as file:
            loaded_data = pickle.load(file)

        print(loaded_data)



**tokenizer_config.json:**

- This file contains the configuration parameters of the BERT tokenizer, such as the tokenizer type, special tokens, vocabulary size, etc.
Example content:


          {
            "max_len": 512,    This indicates the maximum sequence length that the tokenizer will process. Sequences longer than this length will be truncated, and sequences shorter than this length will be padded to reach this length.
            "model_type": "bert",
            "pad_token_id": 0,    This is the token ID used for padding sequences. When sequences are shorter than the maximum length, they are padded with this token ID to match the maximum length.
            "vocab_size": 30522,
            ...
          }


In [5]:
classifier(text)

[{'label': 'POSITIVE', 'score': 0.9967235922813416}]

In [6]:
# Trying to confuse the model
classifier('The enigmatic behavior of dark matter particles, hypothesized to interact solely through gravitational forces yet evading direct detection, underscores the elusive nature of cosmological mysteries that challenge both theoretical frameworks and observational methodologies. ')

[{'label': 'POSITIVE', 'score': 0.9904271364212036}]

# **NER - Named Entity Recoginition**

In [7]:
import pandas as pd

In [8]:
text1 = "Mustafa is data scientist as COMET Estimating LLC."

In [9]:
ner = pipeline('ner', aggregation_strategy='simple')

No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision f2482bf (https://huggingface.co/dbmdz/bert-large-cased-finetuned-conll03-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/998 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.33G [00:00<?, ?B/s]

Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/60.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

In [10]:
pd.DataFrame(ner(text1))

entity_group     score                  word  start  end
0          PER  0.998827               Mustafa      0    7
1          ORG  0.996573  COMET Estimating LLC     29   49

In the Hugging Face Transformers library, the aggregation_strategy parameter in the Named Entity Recognition (NER) pipeline specifies how multiple predictions for the same entity across different tokens are aggregated into a single entity span.



 Let's illustrate the   **aggregation_strategy='simple'**    parameter in the Named Entity Recognition (NER) pipeline with an example:<br><br>










Suppose we have the following input text:


      "The company Google is headquartered in Mountain View, California."


And let's assume the NER model predicts the following entity labels for each token:

      Token:         The   company  Google   is    headquartered   in    Mountain    View       ,        California   .
      Predicted NER: O     O        B-ORG    O     O              O     B-LOC       I-LOC      O        B-LOC        .


The output of the pipeline would be a list of dictionaries, with each dictionary representing an entity found in the input text. For example:


        [
          {'entity': 'ORG', 'text': 'Google', 'start': 12, 'end': 18},
          {'entity': 'LOC', 'text': 'Mountain View', 'start': 34, 'end': 47},
          {'entity': 'LOC', 'text': 'California', 'start': 53, 'end': 63}
        ]




## **Question Answering**

In [11]:
reader = pipeline('question-answering')
question = 'what does user want ?'
outputs = reader(question=question,context = text)
pd.DataFrame([outputs])

No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 626af31 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.
/opt/conda/lib/python3.10/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

score  start  end   answer
0  0.932237    450  457  freedom

## **Summarization**

In [12]:
summarizer = pipeline('summarization')
outputs = summarizer( text, clean_up_tokenization_spaces=True, max_length=90)
print(outputs[0]['summary_text'])

No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.
/opt/conda/lib/python3.10/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/1.80k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

/opt/conda/lib/python3.10/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()


tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

 The only aspect i see as a flaw is the art with a 9/10 the people seem a little bit deformed or too skinny but once you get use to it is great. The almost perfect serie, recomended to anyone who is a human beeing (nevermind, my dog liked it)


In [13]:
# Define the Own text
Eigenvalue = 'Eigenvalue problems involve finding scalars (eigenvalues) and vectors (eigenvectors) such that multiplying the vector by a matrix equals the eigenvalue times the vector.'

# Load the summarization pipeline with the specified model
summarizer = pipeline('summarization', model='facebook/bart-large-cnn')


# Summarize the text
outputs = summarizer(Eigenvalue, clean_up_tokenization_spaces=True, max_length=90)

# Print the summary text
print(outputs[0]['summary_text'])

config.json:   0%|          | 0.00/1.58k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Your max_length is set to 90, but your input_length is only 40. Since this is a summarization task, where outputs shorter than the input are typically wanted, you might consider decreasing max_length manually, e.g. summarizer('...', max_length=20)


Eigenvalue problems involve finding scalars (eigenvalues) and vectors ( eigenvectors) such that multiplying the vector by a matrix equals the eigenvalue times the vector. Eigenvalue problems occur when a scalar and a vector have different eigenvalues.


In [14]:
# Define the Own text 
Eigenvalue = 'Eigenvalue problems involve finding scalars (eigenvalues) and vectors (eigenvectors) such that multiplying the vector by a matrix equals the eigenvalue times the vector.'

# Load the summarization pipeline with the specified model abnother model
summarizer = pipeline('summarization', model='pszemraj/long-t5-tglobal-base-16384-book-summary')


# Summarize the text
outputs = summarizer(Eigenvalue, clean_up_tokenization_spaces=True, max_length=90)

# Print the summary text
print(outputs[0]['summary_text'])

config.json:   0%|          | 0.00/1.05k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/990M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/2.36k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]

Your max_length is set to 90, but your input_length is only 48. Since this is a summarization task, where outputs shorter than the input are typically wanted, you might consider decreasing max_length manually, e.g. summarizer('...', max_length=24)
/opt/conda/lib/python3.10/site-packages/transformers/modeling_utils.py:1006: FutureWarning: The `device` argument is deprecated and will be removed in v5 of Transformers.
  warnings.warn(


In this chapter, we're introduced to the concept of eigenvalence. It's an important concept because it involves two very different sets of problems: deterministic and non-deterministic.


**clean_up_tokenization_spaces=True:**<br><br>

- When clean_up_tokenization_spaces is set to True, it indicates that the pipeline should remove any extra spaces resulting from tokenization before generating the summary text.<br><br>
- Tokenization of text involves splitting it into individual tokens (words or subwords). Sometimes, tokenization can introduce extra spaces, especially around punctuation marks or special characters.<br><br>
- Setting clean_up_tokenization_spaces=True ensures that these extra spaces are removed from the input text before it is summarized, resulting in cleaner and more readable summaries.<br><br>


**max_length=90:**<br><br>

- The max_length parameter specifies the maximum length (in tokens) of the generated summary text.<br><br>
- When summarizing text, it's often desirable to limit the length of the summary to make it concise and focused.<br><br>
- By setting max_length=90, you're instructing the summarization pipeline to generate a summary with a maximum length of 90 tokens. If the summary exceeds this length, it will be truncated to meet the specified limit.<br><br>
- Limiting the length of the summary can help control the amount of information presented to the reader and ensure that the summary remains informative and easy to comprehend.

## **Translation**

In [15]:
translate = pipeline('translation_en_to_de', model='Helsinki-NLP/opus-mt-en-de')
output = translate(text,clean_up_tokenization_spaces=True,min_length=120)
print(output[0]['translation_text'])

config.json:   0%|          | 0.00/1.33k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/298M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/293 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/42.0 [00:00<?, ?B/s]

source.spm:   0%|          | 0.00/768k [00:00<?, ?B/s]

target.spm:   0%|          | 0.00/797k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.27M [00:00<?, ?B/s]

/opt/conda/lib/python3.10/site-packages/transformers/models/marian/tokenization_marian.py:175: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")


Code Geass ist eine jener Serien, die jeder recomends und sobald Sie es sehen Sie wissen, warum. Mechas, Krieg, Romantik, Komödie, Rivalität, unerwartete Twsits, Explotionen, Mindgames und mehr mechas! was sonst wollen Sie! xD es völlig verdient seinen Platz als einer der besten Animes jemals gemacht. der einzige Aspekt, den ich als Fehler sehe, ist die Kunst mit einem 9/10 die Menschen scheinen ein wenig verformt oder zu dünn, aber wenn Sie es nutzen ist toll!.Der Benutzer will Freiheit. und es hat bei weitem die beste Geschichte, die ich je gesehen habe, Sie nur an sie von der zweiten Folge und Cant Stop bis zum Ende. fast perfekt serie, recomended zu jedem, der ein Mensch ist Bienen (Nevermind, mein Hund mochte es)


## **TEXT CLASSIFICATION WITH HUGGINGFACE**

In [16]:
!pip install --upgrade pip
!pip install transformers
!pip install datasets
!pip install sentencepiece

/opt/conda/lib/python3.10/pty.py:89: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid, fd = os.forkpty()


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 27.3 MB/s eta 0:00:00
  Attempting uninstall: pip
    Found existing installation: pip 23.3.2
    Uninstalling pip-23.3.2:
      Successfully uninstalled pip-23.3.2


# Datasets

In [17]:
# Emotion detector
from datasets import list_datasets
all_datasets = list_datasets()
print(f"There are total of {len(all_datasets)} in the hub")
print(f"Some example datasets are {all_datasets[:5]}")

/tmp/ipykernel_25/979804645.py:3: FutureWarning: list_datasets is deprecated and will be removed in the next major version of datasets. Use 'huggingface_hub.list_datasets' instead.
  all_datasets = list_datasets()


There are total of 169039 in the hub
Some example datasets are ['amirveyseh/acronym_identification', 'ade-benchmark-corpus/ade_corpus_v2', 'UCLNLP/adversarial_qa', 'Yale-LILY/aeslc', 'nwu-ctext/afrikaans_ner_corpus']


In [18]:
from datasets import load_dataset
emotion_dataset = load_dataset('emotion')
emotion_dataset

/opt/conda/lib/python3.10/site-packages/datasets/load.py:1491: FutureWarning: The repository for emotion contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/emotion
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(


Generating train split:   0%|          | 0/16000 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/2000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/2000 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 16000
    })
    validation: Dataset({
        features: ['text', 'label'],
        num_rows: 2000
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 2000
    })
})

In [19]:
train_dataset = emotion_dataset['train']
test_dataset = emotion_dataset['test']
train_dataset, test_dataset

(Dataset({
     features: ['text', 'label'],
     num_rows: 16000
 }),
 Dataset({
     features: ['text', 'label'],
     num_rows: 2000
 }))

In [20]:
train_dataset[0],test_dataset[0]

({'text': 'i didnt feel humiliated', 'label': 0},
 {'text': 'im feeling rather rotten so im not very ambitious right now',
  'label': 0})

In [21]:
import pandas as pd
df = pd.DataFrame(train_dataset)
print(df.shape)
df.head(10)

(16000, 2)


text  label
0                            i didnt feel humiliated      0
1  i can go from feeling so hopeless to so damned...      0
2   im grabbing a minute to post i feel greedy wrong      3
3  i am ever feeling nostalgic about the fireplac...      2
4                               i am feeling grouchy      3
5  ive been feeling a little burdened lately wasn...      0
6  ive been taking or milligrams or times recomme...      5
7  i feel as confused about life as a teenager or...      4
8  i have been with petronas for years i feel tha...      1
9                                i feel romantic too      2

In [22]:
train_dataset.column_names

['text', 'label']

In [23]:
train_dataset.features

{'text': Value(dtype='string', id=None),
 'label': ClassLabel(names=['sadness', 'joy', 'love', 'anger', 'fear', 'surprise'], id=None)}

In [24]:
train_dataset.features['label'].names

['sadness', 'joy', 'love', 'anger', 'fear', 'surprise']

In [25]:
!wget https://raw.githubusercontent.com/nachikethmurthy/Source-Code-Dataset-for-Machine-Learning-using-Python/main/Data/sentiment_train

/opt/conda/lib/python3.10/pty.py:89: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid, fd = os.forkpty()


--2024-06-26 15:26:21--  https://raw.githubusercontent.com/nachikethmurthy/Source-Code-Dataset-for-Machine-Learning-using-Python/main/Data/sentiment_train
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 447555 (437K) [text/plain]
Saving to: 'sentiment_train'

sentiment_train     100%[===================>] 437.07K  --.-KB/s    in 0.03s   

2024-06-26 15:26:21 (14.2 MB/s) - 'sentiment_train' saved [447555/447555]



In [26]:
# load the dataset
sentiment_train = load_dataset('csv',data_files = "sentiment_train",sep='\t')
sentiment_train

Generating train split: 0 examples [00:00, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['sentiment', 'text'],
        num_rows: 6918
    })
})

In [27]:
emotion_dataset

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 16000
    })
    validation: Dataset({
        features: ['text', 'label'],
        num_rows: 2000
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 2000
    })
})

Convert the Hugging Face Dataset object (emotion_dataset) into a pandas DataFrame (df)

In [28]:
emotion_dataset.set_format(type='pandas')
df = emotion_dataset['train'][:]
df.head()

text  label
0                            i didnt feel humiliated      0
1  i can go from feeling so hopeless to so damned...      0
2   im grabbing a minute to post i feel greedy wrong      3
3  i am ever feeling nostalgic about the fireplac...      2
4                               i am feeling grouchy      3

In [29]:
emotion_dataset['train']

Dataset({
    features: ['text', 'label'],
    num_rows: 16000
})

In [30]:
emotion_dataset['train'][:]

text  label
0                                i didnt feel humiliated      0
1      i can go from feeling so hopeless to so damned...      0
2       im grabbing a minute to post i feel greedy wrong      3
3      i am ever feeling nostalgic about the fireplac...      2
4                                   i am feeling grouchy      3
...                                                  ...    ...
15995  i just had a very brief time in the beanbag an...      0
15996  i am now turning and i feel pathetic that i am...      0
15997                     i feel strong and good overall      1
15998  i feel like this was such a rude comment and i...      3
15999  i know a lot but i feel so stupid because i ca...      0

[16000 rows x 2 columns]

In [31]:
print(emotion_dataset['train'].features['label'].int2str(0))
print(emotion_dataset['train'].features['label'].int2str(1))
print(emotion_dataset['train'].features['label'].int2str(2))
emotion_dataset['train'].features['label'].names

sadness
joy
love


['sadness', 'joy', 'love', 'anger', 'fear', 'surprise']

**The int2str()**<br><br>


 method is used to convert integer indices back to string labels. It's particularly useful when dealing with classification tasks where labels are represented as integers but you want to convert them back to their original string representations.

In [32]:
def label_to_str(row):
  return emotion_dataset['train'].features['label'].int2str(row)

df['labels_name'] = df['label'].apply(label_to_str)
df.head()

text  label labels_name
0                            i didnt feel humiliated      0     sadness
1  i can go from feeling so hopeless to so damned...      0     sadness
2   im grabbing a minute to post i feel greedy wrong      3       anger
3  i am ever feeling nostalgic about the fireplac...      2        love
4                               i am feeling grouchy      3       anger

In [33]:
emotion_dataset.reset_format()   #Hugging Face datasets library, such as the DatasetDict

In [34]:
from transformers import AutoTokenizer
model = 'distilbert-base-uncased'
tokenizer = AutoTokenizer.from_pretrained(model)

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

/opt/conda/lib/python3.10/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

In [35]:
text = "My favorite food is Mother's Hand made food"
encoded_text = tokenizer(text)
print(encoded_text)

{'input_ids': [101, 2026, 5440, 2833, 2003, 2388, 1005, 1055, 2192, 2081, 2833, 102], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}


      input_ids: This is a list of token IDs representing the input text after tokenization. Each token in the input text is mapped to a corresponding ID from the model's vocabulary. Here's what each ID represents:

      101: This is the special token [CLS], which stands for "classification" and is added at the beginning of the input sequence.
      2026: The token ID for the word "My".
      5440: The token ID for the word "favorite".
      2833: The token ID for the word "food".
      2003: The token ID for the word "is".
      16137: The token ID for the subword "mas".
      7911: The token ID for the subword "##ala".
      9998: The token ID for the subword "dos".
      2050: The token ID for the subword '##a'
      102: This is another instance of the special token [SEP], indicating the end of the sequence.
      attention_mask: This is a list indicating which tokens in the input sequence should be attended to (have their attention scores calculated) and which ones should be ignored. In this case, all tokens have an attention mask value of 1, indicating that they should all be attended to.

**[CLS] stands for "classification"**


- The special token [CLS] stands for "classification" and is added at the beginning of the input sequence specifically for tasks that involve classification or sequence labeling, such as sentiment analysis, text classification, or named entity recognition. Its main purpose is to provide a representation of the entire input sequence that can be used for classification tasks.

- When training a model for classification tasks using techniques like BERT, the model typically receives a single vector representation of the entire input sequence. The [CLS] token's position is chosen because it always appears in the same position in every input sequence, making it a consistent representation to use for classification purposes<br><br>


**[SEP] Token (ID 102):**

- The [SEP] token is indeed a separator token used to mark the end of a sequence or to separate segments within a sequence.
- It's added at the end of each input sequence to indicate the end of the sequence.
- In tasks like sentence classification or question answering, where there's only one input sequence, a single [SEP] token is added at the end to denote the end of that sequence.





In [36]:
type(encoded_text)

transformers.tokenization_utils_base.BatchEncoding

**transformers:**

This is the namespace or package where the class tokenization_utils_base is defined. In the context of Hugging Face's Transformers library, this is where the tokenization-related utilities are implemented.<br><br>

**tokenization_utils_base:**


This is a module within the Transformers library that contains base classes and utilities for tokenization. It provides the foundation for various tokenization-related functionalities.<br><br>

**BatchEncoding:**

This is the specific class or type of object returned by the tokenizer when encoding a batch of input sequences. It represents a batch of encoded sequences and typically contains attributes such as input_ids, attention_mask, and other relevant information about the encoded batch.

In [37]:
tokens = tokenizer.convert_ids_to_tokens(encoded_text.input_ids)
print(tokens)

['[CLS]', 'my', 'favorite', 'food', 'is', 'mother', "'", 's', 'hand', 'made', 'food', '[SEP]']


In [38]:
tokenizer.convert_tokens_to_string(tokens)

"[CLS] my favorite food is mother ' s hand made food [SEP]"

In [39]:
tokenizer.vocab_size

30522

In [40]:
tokenizer.model_max_length

512

In [41]:
tokenizer.model_input_names

['input_ids', 'attention_mask']

## **TOKENIZATION ON DATASET**

In [42]:
emotion_dataset['train'][:2]

{'text': ['i didnt feel humiliated',
  'i can go from feeling so hopeless to so damned hopeful just from being around someone who cares and is awake'],
 'label': [0, 0]}

In [43]:
def tokenization(batch):
  return tokenizer(batch['text'], padding=True, truncation=True)

tokenization(emotion_dataset['train'][:2])

{'input_ids': [[101, 1045, 2134, 2102, 2514, 26608, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [101, 1045, 2064, 2175, 2013, 3110, 2061, 20625, 2000, 2061, 9636, 17772, 2074, 2013, 2108, 2105, 2619, 2040, 14977, 1998, 2003, 8300, 102]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}

        {'input_ids': [[101, 1045, 2134, 2102, 2514, 26608, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                      [101, 1045, 2064, 2175, 2013, 3110, 2061, 20625, 2000, 2061, 9636, 17772, 2074, 2013, 2108, 2105, 2619, 2040, 14977, 1998, 2003, 8300, 102]],
        'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}


In [44]:
emotions_encoded = emotion_dataset.map(tokenization, batched=True, batch_size=None)
emotions_encoded

Map:   0%|          | 0/16000 [00:00<?, ? examples/s]

Map:   0%|          | 0/2000 [00:00<?, ? examples/s]

Map:   0%|          | 0/2000 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 16000
    })
    validation: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 2000
    })
    test: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 2000
    })
})

# **HuggingFace Models**

**BitsAndBytes:**

The bitsandbytes library is a Python library developed by OpenAI for efficient data serialization and deserialization.<br><br>


Serialization and Deserialization
=====================================

### What is Serialization?

Serialization is the process of converting an object's state into a format that can be written to a file, sent over a network, or stored in a database. This allows the object to be reconstructed later in the same or a different programming language.

### What is Deserialization?

Deserialization is the process of reconstructing an object from a serialized format. It's the reverse of serialization.

### Key differences:

|  | Serialization | Deserialization |
| --- | --- | --- |
| **Direction** | Object → Serialized format | Serialized format → Object |
| **Purpose** | Convert object to a storable format | Reconstruct object from stored format |
| **Output** | Serialized data (e.g., JSON, binary) | Original object |

**Here's how it works:**

- Original List: [10, 20, 30, 40, 50]<br><br>
- Serialized Form: 10 20 30 40 50<br><br>
- In the serialized form, each number is represented as a string of characters separated by spaces. This representation may look shorter compared to the original list, but in terms of memory or storage size, it may not necessarily be shorter.<br><br>

- The primary benefit of serialization with bitsandbytes is not to make the data physically shorter but to convert it into a format that is easier to handle, transmit, or store, especially when dealing with complex data structures or when interoperating with other systems.<br><br>



**Accelerate:**<br><br>

- Suppose you're training a deep learning model to recognize images. This process involves lots of complex calculations.<br><br>
- With accelerate, you can harness the power of specialized hardware like GPUs (Graphics Processing Units) to speed up these calculations.<br><br>
- For example, if it takes 10 hours to train your model on a regular computer, accelerate might reduce it to just 2 hours on a GPU.<br><br>
- Similarly, when you're running your trained model to make predictions, accelerate helps make those predictions faster, which is crucial for real-time applications like video processing or natural language understanding.<br><br>
In summary, bitsandbytes helps with data compression and decompression, while accelerate boosts the performance of deep learning tasks by leveraging specialized hardware.














In [45]:
!pip install accelerate

/opt/conda/lib/python3.10/pty.py:89: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid, fd = os.forkpty()


# Mixtral-8x22B-v0.1

      "8x22" indicates that the model consists of 8 layers of 22 blocks each.
      "B" typically represents billion parameters, suggesting that the model has a large number of parameters.
      Now, let's discuss the concept of a Mixture of Experts (MoE) and how it's trained:

      Mixture of Experts (MoE):

      A Mixture of Experts is a neural network architecture that consists of multiple subnetworks called "experts" working in parallel, along with a gating mechanism that learns to dynamically select which expert to use for a given input. Each expert specializes in different aspects of the data, and the gating mechanism determines the relevance or importance of each expert for a particular input.

      In the context of language modeling, each expert might specialize in generating specific types of text, such as formal language, informal language, technical jargon, etc.

      Training Process:

      Expert Training: Initially, each expert in the MoE is trained independently on the data. This allows each expert to specialize in different aspects of the data.

      Gating Network Training: Simultaneously, the gating network is trained to learn how to dynamically combine the outputs of the experts based on the input data. The gating network's goal is to select the most relevant expert for each input.

      Joint Training: After training the experts and the gating network separately, the entire MoE model is fine-tuned jointly. During this joint training, the parameters of both the experts and the gating network are updated together to optimize the overall performance of the model.

      Example:

      Consider a language generation task where the input is a prompt, and the model is tasked with generating a coherent continuation of the prompt.

      Expert Networks: The MoE might consist of experts trained on different genres of text, such as news articles, fiction, scientific papers, etc.

      Gating Mechanism: The gating network learns to identify the genre or style of the input prompt and dynamically selects the expert best suited to generate text in that style.

      Training: During training, the MoE is fed a variety of prompts from different genres, and the parameters of both the experts and the gating network are adjusted to minimize the generation loss across all inputs.

      Overall, the Mixture of Experts architecture enables the model to leverage the strengths of multiple specialized subnetworks, resulting in improved performance and versatility across different types of data.








# Mixture of Experts Model Working with example

      Imagine you're planning a big event, like a party, and you need to organize various activities for different groups of guests. You decide to hire a team of experts, each specializing in a different aspect of event planning:

      Food Expert:
      This expert knows all about cooking and catering delicious meals.

      Music Expert:
      This expert specializes in creating playlists and setting up sound systems for entertainment.

      Decoration Expert:
      This expert is skilled at designing and decorating the venue to create a festive atmosphere.

      Entertainment Expert:
      This expert arranges games, performances, and other activities to keep guests entertained.

      Now, here's how the MoE architecture works:

      Expert Training:
      Each expert receives specialized training in their respective area. The food expert learns recipes, the music expert curates playlists, the decoration expert hones their design skills, and the entertainment expert plans fun activities.

      Gating Mechanism:
      You, as the event organizer, act as the gating mechanism. Based on the preferences and interests of your guests, you dynamically select which expert to consult for each aspect of the event. For example, if you have guests who love dancing, you might prioritize the music expert and choose lively playlists.

      Joint Training:
      As you plan the event, you interact with the experts and fine-tune the details. You might adjust the menu based on dietary preferences, tweak the playlist to suit different moods, and personalize the decorations to match the theme. This collaborative process optimizes the overall event experience.

      In summary, the MoE model architecture combines the expertise of specialized subnetworks (the experts) with a dynamic selection mechanism (the gating mechanism) to optimize performance across different tasks or domains. By leveraging the strengths of multiple experts, the MoE can adapt to diverse inputs and produce high-quality outputs tailored to specific needs.








In [46]:
# Before runing this cell make sure you have enough Ram

# from transformers import AutoModelForCausalLM, AutoTokenizer

# model_id = "mistral-community/Mixtral-8x22B-v0.1"
# tokenizer = AutoTokenizer.from_pretrained(model_id)

# model = AutoModelForCausalLM.from_pretrained(model_id)

# text = "Hello my name is"
# inputs = tokenizer(text, return_tensors="pt")

# outputs = model.generate(**inputs, max_new_tokens=20)
# print(tokenizer.decode(outputs[0], skip_special_tokens=True))


**AutoModelForCausalLM and AutoTokenizer**<br><br>
These are classes provided by the Hugging Face transformers library. They are used to simplify the process of loading and using pre-trained language models and tokenizers.<br><br>

**AutoModelForCausalLM:**

- This class is used to load pre-trained language models that are specifically designed for causal language modeling tasks.<br><br>
- Causal language modeling is a type of language modeling where the model predicts the next token in a sequence given the preceding tokens. It is commonly used for tasks like text generation and completion.<br><br>
- By using AutoModelForCausalLM, you can easily load a pre-trained model suitable for causal language modeling without worrying about the specific model architecture or configuration.<br><br>


**AutoTokenizer:**<br><br>

- This class is used to load pre-trained tokenizers that are compatible with specific language models.<br><br>
- Tokenizers are used to preprocess text data into tokens that can be fed into a language model.<br><br>
- The AutoTokenizer class automatically selects the appropriate tokenizer based on the specified model_id.

1.  **Table Question Answering from Tasks**




In [47]:
from transformers import pipeline
import pandas as pd

# prepare table + question
data = {"Actors": ["Brad Pitt", "Leonardo Di Caprio", "George Clooney"], "Number of movies": ["87", "53", "69"]}
table = pd.DataFrame.from_dict(data)
question = "how many movies does Leonardo Di Caprio have?"

# pipeline model
# Note: you must to install torch-scatter first.
tqa = pipeline(task="table-question-answering", model="google/tapas-large-finetuned-wtq")

# result

print(tqa(table=table, query=question)['cells'][0])
#53

config.json:   0%|          | 0.00/1.66k [00:00<?, ?B/s]

TAPAS models are not usable since `tensorflow_probability` can't be loaded. It seems you have `tensorflow_probability` installed with the wrong tensorflow version. Please try to reinstall it following the instructions here: https://github.com/tensorflow/probability.


model.safetensors:   0%|          | 0.00/1.35G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/490 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/262k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/154 [00:00<?, ?B/s]

/opt/conda/lib/python3.10/site-packages/transformers/models/tapas/tokenization_tapas.py:2674: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  text = normalize_for_match(row[col_index].text)
/opt/conda/lib/python3.10/site-packages/transformers/models/tapas/tokenization_tapas.py:1473: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  cell = row[col_index]


53


In [48]:
print(tqa(table=table, query="bradd pitt movies")['cells'][0])

87


2.  **Document-Question-Answering from Tasks**

In [49]:
from transformers import pipeline
from PIL import Image
import requests
from io import BytesIO

# Initialize the pipeline
pipe = pipeline("document-question-answering", model="naver-clova-ix/donut-base-finetuned-docvqa")

# Define the question
question = "What is the purchase outer box?"

# Load the image from a URL
url = "https://th.bing.com/th/id/R.e1a65119dd06af8add659ff4753d00d4?rik=BzjY6TpT7iOdAg&pid=ImgRaw&r=0"
response = requests.get(url)
image = Image.open(BytesIO(response.content))

# Perform document question answering
answer = pipe(image=image, question=question)

# Print the answer
print(answer)

config.json:   0%|          | 0.00/4.74k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/803M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/535 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/1.30M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/4.01M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/478 [00:00<?, ?B/s]

preprocessor_config.json:   0%|          | 0.00/359 [00:00<?, ?B/s]

/opt/conda/lib/python3.10/site-packages/transformers/generation/utils.py:1168: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(


[{'answer': 'deep learning'}]


In [50]:
# Perform document question answering
answer = pipe(image=image, question= 'How many part are there in the pitcher' )

# Print the answer
print(answer)

[{'answer': 'deep'}]


## **Evaluation Metrics**

# Bert Score

BERTScore is a metric used for evaluating the quality of text generated by language models. It measures how similar the generated text (predictions) is to the reference text.<br><br>

Here's a simpler breakdown:<br><br>

**What it measures:**<br><br>

- BERTScore calculates a score for each token in the generated text (predictions) by comparing it to each token in the reference text. It uses pre-trained language models like BERT to understand the context of the words.<br><br>

**How it works:**<br><br>

- BERTScore looks at each word in the generated text and finds the most similar word in the reference text using cosine similarity. It then calculates precision, recall, and F1 score based on these similarities.

**What it tells us:**<br><br>

- The precision score tells us how many words in the generated text match words in the reference text. The recall score tells us how many words in the reference text are captured by the generated text. The F1 score is a combination of precision and recall, providing an overall measure of similarity.<br><br>

**Usage:** <br><br>
- BERTScore can be used to compare different language generation models, evaluate the quality of machine translation, text summarization, and other natural language processing tasks.

In summary, BERTScore is a metric that measures the similarity between generated text and reference text, providing insight into the quality of language generation models. It's a useful tool for evaluating and comparing different text generation techniques.








In [51]:
# Install the bert_score library
!pip install bert_score

# Import the necessary functions
from bert_score import score

# Prepare your predictions and references
predictions = ["hello world", "general kenobi"]
references = ["hello world", "general kenobi"]

# Compute BERTScore specifying the language (English)
precision, recall, f1 = score(predictions, references, lang="en")

# Print the results
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)


/opt/conda/lib/python3.10/pty.py:89: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid, fd = os.forkpty()


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 3.2 MB/s eta 0:00:00


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

/opt/conda/lib/python3.10/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/482 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.42G [00:00<?, ?B/s]

Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Precision: tensor([1., 1.])
Recall: tensor([1., 1.])
F1 Score: tensor([1., 1.])


In [52]:
!pip install evaluate

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 3.2 MB/s eta 0:00:00


In [53]:
from huggingface_hub import notebook_login

notebook_login()

In [54]:
from evaluate import load
bertscore = load("bertscore")
predictions = ["hello there", "general kenobi"]
references = ["hello there", "general kenobi"]
results = bertscore.compute(predictions=predictions, references=references, lang="en")

Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


# BLEU (Bilingual Evaluation Understudy):

**What it measures:**<br><br>

- BLEU is a metric used to evaluate the quality of machine-translated text by comparing it to human translations. It measures how closely the machine-generated text matches the human reference translations.<br><br>

**How it works:**<br><br>

      Segmentation:
      BLEU first divides the candidate translation (machine-generated text) and reference translations (human-generated text) into segments, usually sentences.

      N-gram matching:
      It then compares each segment of the candidate translation to the corresponding segments in the reference translations. BLEU considers the overlap of n-grams (contiguous sequences of n words) between the candidate and reference translations.

      Precision calculation:
      For each n-gram size, BLEU calculates precision, which measures the proportion of n-grams in the candidate translation that appear in the reference translations.

      Brevity penalty:
      BLEU also penalizes translations that are shorter than the reference translations to prevent shorter translations from receiving artificially high scores.

      BLEU score:
      Finally, BLEU combines the precision scores for different n-gram sizes using a geometric mean and applies the brevity penalty to compute the overall BLEU score

**Scoring range:**<br><br>
- BLEU scores range from 0 to 1, where a score closer to 1 indicates a better match between the machine-generated and human-translated texts. A score of 1 means the machine translation is identical to a human translation.<br><br>

**Usage:**<br><br>
- BLEU is commonly used for evaluating machine translation systems. It helps researchers and developers assess the quality of their translation models and make improvements.<br><br>

In summary, BLEU provides a quantitative measure of the quality of machine translations by comparing them to human translations. Higher BLEU scores indicate better quality translations. It's a widely-used and cost-effective metric for assessing translation performance.







In [55]:
import nltk
from nltk.translate.bleu_score import sentence_bleu

# Define candidate and reference sentences
candidate = "hello world"
reference = ["hello world", "general kenobi"]

# Tokenize candidate and reference sentences
candidate_tokens = candidate.split()
reference_tokens = [ref.split() for ref in reference]

# Calculate BLEU score
bleu_score = sentence_bleu(reference_tokens, candidate_tokens)

# Print BLEU score
print("BLEU Score:", bleu_score)

BLEU Score: 1.0


/opt/conda/lib/python3.10/site-packages/nltk/translate/bleu_score.py:490: UserWarning: 
Corpus/Sentence contains 0 counts of 3-gram overlaps.
BLEU scores might be undesirable; use SmoothingFunction().
  warnings.warn(_msg)


# GLUE (General Language Understanding Evaluation)

- GLUE, or General Language Understanding Evaluation, is a test used to measure how well a machine understands human language. It includes different datasets to evaluate various aspects of language understanding.<br><br>

To use GLUE:<br><br>

**Load the relevant metric:**<br><br>

- Depending on which part of GLUE you're testing against, you load the corresponding metric.<br><br>

**Calculate the metric:**<br><br>

- Once loaded, provide your model's predictions and the reference answers. The metric then computes scores like accuracy or correlation.

**For example, if you're testing on a subset called MRPC, which checks accuracy and F1 score:**

      from evaluate import load

      # Load the GLUE metric for the MRPC subset
      glue_metric = load('glue', 'mrpc')

      # Prepare your predictions and references
      predictions = [0, 1]
      references = [0, 1]

      # Compute the GLUE score
      results = glue_metric.compute(predictions=predictions, references=references)

      # Print the results
      print(results)


**For STSB, which checks Pearson and Spearman correlations:**


        from evaluate import load

        # Load the GLUE metric for the STSB subset
        glue_metric = load('glue', 'stsb')

        # Prepare your predictions and references
        predictions = [-10., -11., -12., -13., -14., -15.]
        references = [0., 1., 2., 3., 4., 5.]

        # Compute the GLUE score
        results = glue_metric.compute(predictions=predictions, references=references)

        # Print the results
        print(results)


# PERPEXILITY

      Imagine you have a language model, like GPT-2, which has been trained on a large corpus of text data. Perplexity measures how well this model predicts a sequence of words.

      Let's say we have a simple sentence: "The cat sat on the mat." Now, the model tries to predict each word in this sentence based on the context of the previous words.

      Word Prediction: For each word in the sentence, the model predicts the probability distribution of the next word. For example:

      For "The", the model predicts the probability of "cat", "dog", "house", etc.
      For "cat", it predicts the probability of "sat", "jumped", "ate", etc.
      And so on for each word in the sentence.
      Perplexity Calculation: Perplexity calculates how surprised the model is by the actual words in the sequence. If the model is very confident in its predictions, the perplexity is low. If it's unsure and assigns similar probabilities to multiple words, the perplexity is high.

      Here's how it works:

      For each word in the sequence, the model assigns a probability to the actual word that appears next.
      Perplexity is the geometric mean of the inverse probabilities of the actual words. In simpler terms, it's like asking: "How surprised was the model by the actual sequence of words?"
      Let's say the model's predictions for our sentence are as follows:

      "The" -> "cat" (with probability 0.8)
      "cat" -> "sat" (with probability 0.7)
      "sat" -> "on" (with probability 0.6)
      "on" -> "the" (with probability 0.9)
      "the" -> "mat" (with probability 0.8)
      To calculate perplexity:

      Inverse the probabilities: 1/0.8, 1/0.7, 1/0.6, 1/0.9, 1/0.8.
      Calculate the geometric mean of these inverse probabilities.
      The result is the perplexity score.
      The lower the perplexity, the better the model's predictions match the actual sequence of words. In our example, if the perplexity is around 1, it means the model was very confident and not surprised by the actual sequence. If it's higher, it means the model was more uncertain about its predictions.








| Date (YYYY-MM-DD) | Version | Changed By           | Change Description         |
| ----------------- | ------- | -------------------- | ---------------------------|
| 2024-05-24        | 0.5     | Mustafa Shoukat      | Complete Exercises'content |
| 2024-06-25        | 0.5     | Maryam  Shoukat      | Review and edit            |


<div style="border-radius: 20px; border: 2px solid #64B5F6; padding: 15px; background-color: #000000; color: white; text-align: center; box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.4), 0px 6px 20px rgba(0, 0, 0, 0.19); transform: perspective(1000px) rotateX(5deg) rotateY(-5deg); transition: transform 0.5s ease-in-out;">
    <h1 style="color: white; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.4); font-weight: bold; margin-bottom: 10px; font-size: 32px;">
        I appreciate your Learning form this notebook
    </h1>
    <h2 style="color: white; text-shadow: 1px 1px 2px rgba(0, 0, 0, 0.3); font-weight: normal; margin-bottom: 10px; font-size: 24px;">
       📓 Your feedback will be important as I try more advanced techniques.🚀
    </h2>
    <h2 style="color: white; text-shadow: 1px 1px 2px rgba(0, 0, 0, 0.3); font-weight: normal; margin-bottom: 10px; font-size: 24px;">
        Suggestions are always welcome!
    </h2>
    <h2 style="color: white; text-shadow: 1px 1px 2px rgba(0, 0, 0, 0.3); font-weight: normal; font-size: 24px;">
        Let's keep pushing forward together!
    </h2>
</div>
